### ISSUE TITLE
A too-large rho value causes a failure, but maybe it shouldn't

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#2070]
#### Issue description

While working with #2048, I noticed that if `rho` is longer than the sequence length, an error is given:

```

error: Cube::slice(): index out of bounds
terminate called after throwing an instance of 'std::logic_error'
  what():  Cube::slice(): index out of bounds
Aborted
```

However, if `rho` is longer than the sequence length, I think we should instead handle this gracefully, and only perform a number of BPTT steps equal to the sequence length.  (I am not 100% certain on that, but at a glance, that seems like it should be the right thing to do.)

#### Your environment

 * version of mlpack: git master
 * operating system: Debian :penguin:
 * compiler: g++ 8.3.0
 * version of dependencies (Boost/Armadillo): Boost 1.67.0, Armadillo 9.600.5
 * any other environment information you think is relevant: This should be reproducible anywhere.

#### Steps to reproduce

You can compile the adapted example from #2048:

```
#include <mlpack/core.hpp>
#include <mlpack/methods/ann/rnn.hpp>
#include <mlpack/methods/ann/ffn.hpp>
#include <mlpack/methods/ann/layer/lstm.hpp>
#include <mlpack/methods/ann/layer/dropout.hpp>

#include <vector>
#include <string>

int main(int argc, char *argv[])
{
  using namespace mlpack;
  using namespace mlpack::data;
  using namespace mlpack::tree;
  using namespace mlpack::cv;
  using namespace mlpack::ann;
  using namespace arma;

  // SUCCEEDS if maxLineLength = 17.
  // FAILS if maxLineLength = 18.
  const int maxLineLength = 18;
  const int hiddenSize = 128;
  const int numLetters = 256;

  using MatType = cube;
  std::vector<std::string> trainingData;
  trainingData.push_back(std::string("THIS IS THE INPUT"));

  //This is the orignal network that I used (FFN) but then I also tried a RNN
  RNN<> rnn(maxLineLength);
  rnn.Add<IdentityLayer<>>();
  rnn.Add<LSTM<>>(numLetters, hiddenSize, maxLineLength);
  rnn.Add<Dropout<>>(0.1);
  rnn.Add<Linear<>>(hiddenSize, numLetters);

  const auto makeInput = [](const char *line) -> MatType {
    const auto strLen = strlen(line);
    // rows: number of dimensions
    // cols: number of sequences/points
    // slices: number of steps in sequences
    MatType result(numLetters, 1, strLen, fill::zeros);
    for(int i = 0; i < strLen; ++i)
    {
      const auto letter = line[i];
      result.at(static_cast<uword>(letter), 0, i) = 1.0;
    }
    return result;
  };

  const auto makeTarget = [] (const char *line) -> MatType {
    const auto strLen = strlen(line);
    // responses for NegativeLogLikelihood should be
    // non-one-hot-encoded class IDs (from 1 to num_classes)
    cube result(1, 1, strLen, fill::zeros);
    // the response is the *next* letter in the sequence
    for(int i = 0; i < strLen - 1; ++i)
    {
      const auto letter = line[i + 1];
      result.at(0, 0, i) = static_cast<uword>(letter) + 1.0;
    }
    // the final response is empty, so we set it to class 0
    result.at(0, 0, strLen - 1) = 1.0;
    return result;
  };

  std::vector<cube> inputs(trainingData.size());
  std::vector<cube> targets(trainingData.size());
  for(int i = 0; i < trainingData.size(); ++i)
  {
    inputs[i] = makeInput(trainingData[i].c_str());
    targets[i] = makeTarget(trainingData[i].c_str());
  }

  ens::SGD<> sgd(0.01, 1, 100 /* only 100 maximum iterations, just to see it work */);
  rnn.Train(inputs[0], targets[0], sgd);
  return 0;
}
```

I compiled on my system in my mlpack build directory with

```
$ g++ -g -O0 -o test ../test.cpp -Iinclude/ -Llib/ -lmlpack -larmadillo -I deps/stb/ -I deps/ensmallen-2.10.3/include/ -fopenmp
```

#### Expected behavior

When `maxLineLength` is set to anything greater than 17, training should proceed without failure.

#### Actual behavior

With `maxLineLength` at 18, we get

```
$ ./test 

error: Cube::slice(): index out of bounds
terminate called after throwing an instance of 'std::logic_error'
  what():  Cube::slice(): index out of bounds
Aborted
```

#### Notes for helping out

If you're interested in working on this issue, here are a few good suggestions for how to get started.

1. Have some understanding of what backpropagation through time is (BPTT); if you aren't already familiar, it would be helpful to know the algorithm, so that the code can be more easily understood.
2. Read and understand the `RNN<>` class in `src/mlpack/methods/ann/`; specifically, the use of the `rho` variable.
3. Try tracing the error with `gdb` to understand exactly what is going on.
4. Adapt the code in the `RNN` class to specifically handle the case when `rho` is greater than the sequence length.

In the end, it's likely to be a lot of time reading and understanding the code, but the patch itself will probably be fairly simple.

There's no need to claim the issue---just work on it, and once you have a working solution, open a PR and someone will review it when they have a chance. :)

**Comment by himanshupathak21061998** [comment]
@rcurtin I want to work on this issue

**Comment by rcurtin** [comment]
Solved via #2102.  Thanks @himanshupathak21061998!
